File: javatests/com/google/protobuf/contrib/j2cl/integration/ByteStringFieldsTest.java
/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.protobuf.contrib.j2cl.integration;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import com.google.protobuf.ByteString;
import com.google.protobuf.contrib.j2cl.protos.Accessor.TestProto;
import java.util.Arrays;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class ByteStringFieldsTest {

  private static final ByteString TEST_STRING = ByteString.copyFrom(new byte[] {1, 2, 3, 4});
  private static final ByteString DEFAULT_VALUE = ByteString.copyFromUtf8("a bytey default");

  @Test
  public void testOptionalFieldNoDefault_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().getOptionalBytes()).isEqualTo(ByteString.EMPTY);
    assertThat(TestProto.getDefaultInstance().hasOptionalBytes()).isFalse();
  }

  @Test
  public void testOptionalFieldNoDefault_setTestBytes() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalBytes(TEST_STRING);
    assertThat(builder.hasOptionalBytes()).isTrue();
    assertThat(TEST_STRING.equals(builder.getOptionalBytes())).isTrue();

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalBytes()).isTrue();
    assertThat(TEST_STRING.equals(proto.getOptionalBytes())).isTrue();
  }

  @Test
  public void testOptionalFieldNoDefault_setEmptyBytes() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalBytes(ByteString.EMPTY);
    assertThat(builder.hasOptionalBytes()).isTrue();
    assertThat(builder.getOptionalBytes()).isEqualTo(ByteString.EMPTY);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalBytes()).isTrue();
    assertThat(proto.getOptionalBytes()).isEqualTo(ByteString.EMPTY);
  }

  @Test
  public void testOptionalFieldNoDefault_clear() {
    TestProto startProto = TestProto.newBuilder().setOptionalBytes(TEST_STRING).build();

    TestProto.Builder builder = startProto.toBuilder();
    builder.clearOptionalBytes();
    assertThat(builder.hasOptionalBytes()).isFalse();

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalBytes()).isFalse();
  }

  @Test
  public void testOptionalFieldWithDefault_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().getOptionalBytesWithDefault())
        .isEqualTo(DEFAULT_VALUE);

    assertThat(TestProto.getDefaultInstance().hasOptionalBytesWithDefault()).isFalse();
  }

  @Test
  public void testOptionalFieldWithDefault_setDefault() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalBytesWithDefault(DEFAULT_VALUE);
    assertThat(builder.hasOptionalBytesWithDefault()).isTrue();
    assertThat(builder.getOptionalBytesWithDefault()).isEqualTo(DEFAULT_VALUE);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalBytesWithDefault()).isTrue();
    assertThat(proto.getOptionalBytesWithDefault()).isEqualTo(DEFAULT_VALUE);
  }

  @Test
  public void testOptionalFieldWithDefault_clear() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalBytesWithDefault(TEST_STRING);
    assertThat(builder.hasOptionalBytesWithDefault()).isTrue();
    assertThat(builder.getOptionalBytesWithDefault()).isEqualTo(TEST_STRING);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalBytesWithDefault()).isTrue();
    assertThat(proto.getOptionalBytesWithDefault()).isEqualTo(TEST_STRING);
  }

  @Test
  public void testRepeatedField_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().getRepeatedBytesCount()).isEqualTo(0);
    assertThat(TestProto.newBuilder().build().getRepeatedBytesCount()).isEqualTo(0);
    assertThrows(Exception.class, () -> TestProto.newBuilder().build().getRepeatedBytes(0));
  }

  @Test
  public void testRepeatedField_add() {
    TestProto.Builder builder =
        TestProto.newBuilder().addRepeatedBytes(TEST_STRING).addRepeatedBytes(ByteString.EMPTY);
    assertThat(builder.getRepeatedBytesCount()).isEqualTo(2);
    assertThat(Arrays.asList(TEST_STRING, ByteString.EMPTY))
        .containsExactlyElementsIn(builder.getRepeatedBytesList())
        .inOrder();

    TestProto proto = builder.build();
    assertThat(proto.getRepeatedBytesCount()).isEqualTo(2);
    assertThat(Arrays.asList(TEST_STRING, ByteString.EMPTY))
        .containsExactlyElementsIn(proto.getRepeatedBytesList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_addAll() {
    TestProto startProto =
        TestProto.newBuilder()
            .addRepeatedBytes(TEST_STRING)
            .addRepeatedBytes(ByteString.EMPTY)
            .build();

    TestProto.Builder builder =
        startProto
            .toBuilder()
            .addAllRepeatedBytes(
                Arrays.asList(
                    ByteString.copyFrom(new byte[] {1, 2, 3}),
                    ByteString.copyFrom(new byte[] {4, 5, 6}),
                    ByteString.copyFrom(new byte[] {7, 8, 9})));

    assertThat(
            Arrays.asList(
                TEST_STRING,
                ByteString.EMPTY,
                ByteString.copyFrom(new byte[] {1, 2, 3}),
                ByteString.copyFrom(new byte[] {4, 5, 6}),
                ByteString.copyFrom(new byte[] {7, 8, 9})))
        .containsExactlyElementsIn(builder.getRepeatedBytesList())
        .inOrder();

    TestProto proto = builder.build();

    assertThat(
            Arrays.asList(
                TEST_STRING,
                ByteString.EMPTY,
                ByteString.copyFrom(new byte[] {1, 2, 3}),
                ByteString.copyFrom(new byte[] {4, 5, 6}),
                ByteString.copyFrom(new byte[] {7, 8, 9})))
        .containsExactlyElementsIn(proto.getRepeatedBytesList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_set() {
    TestProto startProto =
        TestProto.newBuilder()
            .addRepeatedBytes(TEST_STRING)
            .addRepeatedBytes(ByteString.EMPTY)
            .addRepeatedBytes(ByteString.copyFrom(new byte[] {0, 0, 0}))
            .build();

    TestProto.Builder builder =
        startProto.toBuilder().setRepeatedBytes(2, ByteString.copyFrom(new byte[] {3, 3, 3}));

    assertThat(
            Arrays.asList(TEST_STRING, ByteString.EMPTY, ByteString.copyFrom(new byte[] {3, 3, 3})))
        .containsExactlyElementsIn(builder.getRepeatedBytesList())
        .inOrder();

    TestProto proto = builder.build();

    assertThat(
            Arrays.asList(TEST_STRING, ByteString.EMPTY, ByteString.copyFrom(new byte[] {3, 3, 3})))
        .containsExactlyElementsIn(proto.getRepeatedBytesList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_getAndCount() {
    TestProto.Builder builder =
        TestProto.newBuilder()
            .addRepeatedBytes(TEST_STRING)
            .addRepeatedBytes(ByteString.EMPTY)
            .addRepeatedBytes(ByteString.EMPTY);

    assertThat(builder.getRepeatedBytesCount()).isEqualTo(3);
    assertThat(builder.getRepeatedBytes(0)).isEqualTo(TEST_STRING);
    assertThat(builder.getRepeatedBytes(1)).isEqualTo(ByteString.EMPTY);
    assertThat(builder.getRepeatedBytes(2)).isEqualTo(ByteString.EMPTY);
    if (InternalChecks.isCheckIndex()) {
      assertThrows(Exception.class, () -> builder.getRepeatedBytes(3));
    }

    TestProto proto = builder.build();
    assertThat(proto.getRepeatedBytesCount()).isEqualTo(3);
    assertThat(proto.getRepeatedBytes(0)).isEqualTo(TEST_STRING);
    assertThat(proto.getRepeatedBytes(1)).isEqualTo(ByteString.EMPTY);
    assertThat(proto.getRepeatedBytes(2)).isEqualTo(ByteString.EMPTY);

    if (InternalChecks.isCheckIndex()) {
      assertThrows(Exception.class, () -> proto.getRepeatedBytes(3));
    }
  }

  @Test
  public void testRepeatedField_clear() {
    TestProto startProto =
        TestProto.newBuilder()
            .addRepeatedBytes(TEST_STRING)
            .addRepeatedBytes(ByteString.EMPTY)
            .addRepeatedBytes(ByteString.copyFrom(new byte[] {0, 0, 0}))
            .build();

    TestProto.Builder builder = startProto.toBuilder().clearRepeatedBytes();
    assertThat(builder.getRepeatedBytesCount()).isEqualTo(0);
    assertThat(builder.getRepeatedBytesList()).isEmpty();

    TestProto proto = builder.build();
    assertThat(proto.getRepeatedBytesCount()).isEqualTo(0);
    assertThat(proto.getRepeatedBytesCount()).isEqualTo(0);
  }
}

File: javatests/com/google/protobuf/contrib/immutablejs/integration/bytestring_fields_test.js
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

goog.module('proto.im.integration.ByteStringFieldsTest');
goog.setTestOnly();

const ByteString = goog.require('proto.im.ByteString');
const ListView = goog.require('proto.im.ListView');
const TestProto = goog.require('improto.protobuf.contrib.immutablejs.protos.TestProto');
const testSuite = goog.require('goog.testing.testSuite');
const {assertEqualsForProto} = goog.require('proto.im.proto_asserts');
const {isCheckIndex} = goog.require('proto.im.internal.internalChecks');


const TEST_STRING = ByteString.copyFrom([1, 2, 3, 4]);
const DEFAULT_VALUE = ByteString.copyFrom(stringToBytes('a bytey default'));

/**
 * Converts a string into an array of char codes.
 *
 * Only works for ascii strings
 *
 * @param {string }val
 * @return {!Array<number>}
 */
function stringToBytes(val) {
  const result = [];
  for (let i = 0; i < val.length; i++) {
    result.push(val.charCodeAt(i));
  }
  return result;
}

class ByteStringFieldsTest {
  testOptionalFieldNoDefault_defaultInstance() {
    assertEqualsForProto(
        ByteString.EMPTY, TestProto.newBuilder().build().getOptionalBytes());
    assertEqualsForProto(
        false, TestProto.newBuilder().build().hasOptionalBytes());
  }

  testOptionalFieldNoDefault_setTestBytes() {
    const builder = TestProto.newBuilder().setOptionalBytes(TEST_STRING);
    assertEqualsForProto(true, builder.hasOptionalBytes());
    assertEqualsForProto(TEST_STRING, builder.getOptionalBytes());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalBytes());
    assertEqualsForProto(TEST_STRING, proto.getOptionalBytes());
  }

  testOptionalFieldNoDefault_setEmptyBytes() {
    const builder = TestProto.newBuilder().setOptionalBytes(ByteString.EMPTY);
    assertEqualsForProto(true, builder.hasOptionalBytes());
    assertEqualsForProto(ByteString.EMPTY, builder.getOptionalBytes());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalBytes());
    assertEqualsForProto(ByteString.EMPTY, proto.getOptionalBytes());
  }

  testOptionalFieldNoDefault_clear() {
    const startProto =
        TestProto.newBuilder().setOptionalBytes(TEST_STRING).build();

    // Test clear
    const builder = startProto.toBuilder();
    builder.clearOptionalBytes();
    assertEqualsForProto(false, builder.hasOptionalBytes());

    const proto = builder.build();
    assertEqualsForProto(false, proto.hasOptionalBytes());
  }

  testOptionalFieldWithDefault_defaultInstance() {
    assertEqualsForProto(
        DEFAULT_VALUE,
        TestProto.newBuilder().build().getOptionalBytesWithDefault());

    assertEqualsForProto(
        false, TestProto.newBuilder().build().hasOptionalBytesWithDefault());
  }

  testOptionalFieldWithDefault_setDefault() {
    const builder =
        TestProto.newBuilder().setOptionalBytesWithDefault(DEFAULT_VALUE);
    assertEqualsForProto(true, builder.hasOptionalBytesWithDefault());
    assertEqualsForProto(DEFAULT_VALUE, builder.getOptionalBytesWithDefault());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalBytesWithDefault());
    assertEqualsForProto(DEFAULT_VALUE, proto.getOptionalBytesWithDefault());
  }

  testOptionalFieldWithDefault_clear() {
    const builder =
        TestProto.newBuilder().setOptionalBytesWithDefault(TEST_STRING);
    assertEqualsForProto(true, builder.hasOptionalBytesWithDefault());
    assertEqualsForProto(TEST_STRING, builder.getOptionalBytesWithDefault());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalBytesWithDefault());
    assertEqualsForProto(TEST_STRING, proto.getOptionalBytesWithDefault());
  }

  testRepeatedField_defaultInstance() {
    assertEqualsForProto(
        0, TestProto.newBuilder().build().getRepeatedBytesCount());
    assertEqualsForProto(
        0, TestProto.newBuilder().build().getRepeatedBytesList().size());
    if (isCheckIndex()) {
      assertThrows(() => TestProto.newBuilder().build().getRepeatedBytes(0));
    } else {
      assertUndefined(TestProto.newBuilder().build().getRepeatedBytes(0));
    }
  }

  testRepeatedField_add() {
    const builder = TestProto.newBuilder()
                        .addRepeatedBytes(TEST_STRING)
                        .addRepeatedBytes(ByteString.EMPTY);
    assertEqualsForProto(2, builder.getRepeatedBytesCount());
    assertEqualsForProto(
        [TEST_STRING, ByteString.EMPTY],
        builder.getRepeatedBytesList().toArray());

    const proto = builder.build();
    assertEqualsForProto(2, proto.getRepeatedBytesCount());
    assertEqualsForProto(
        [TEST_STRING, ByteString.EMPTY],
        proto.getRepeatedBytesList().toArray());
  }

  testRepeatedField_addAll() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedBytes(TEST_STRING)
                           .addRepeatedBytes(ByteString.EMPTY)
                           .build();

    // test addAll
    const builder = startProto.toBuilder().addAllRepeatedBytes(ListView.copyOf([
      ByteString.copyFrom([1, 2, 3]), ByteString.copyFrom([4, 5, 6]),
      ByteString.copyFrom([7, 8, 9])
    ]));

    assertEqualsForProto(5, builder.getRepeatedBytesCount());
    assertEqualsForProto(
        [
          TEST_STRING, ByteString.EMPTY, ByteString.copyFrom([1, 2, 3]),
          ByteString.copyFrom([4, 5, 6]), ByteString.copyFrom([7, 8, 9])
        ],
        builder.getRepeatedBytesList().toArray());
    assertEqualsForProto(TEST_STRING, builder.getRepeatedBytes(0));
    assertEqualsForProto(ByteString.EMPTY, builder.getRepeatedBytes(1));
    assertEqualsForProto(
        ByteString.copyFrom([1, 2, 3]), builder.getRepeatedBytes(2));
    assertEqualsForProto(
        ByteString.copyFrom([4, 5, 6]), builder.getRepeatedBytes(3));
    assertEqualsForProto(
        ByteString.copyFrom([7, 8, 9]), builder.getRepeatedBytes(4));


    const proto = builder.build();
    assertEqualsForProto(5, proto.getRepeatedBytesCount());
    assertEqualsForProto(
        [
          TEST_STRING, ByteString.EMPTY, ByteString.copyFrom([1, 2, 3]),
          ByteString.copyFrom([4, 5, 6]), ByteString.copyFrom([7, 8, 9])
        ],
        proto.getRepeatedBytesList().toArray());
    assertEqualsForProto(TEST_STRING, proto.getRepeatedBytes(0));
    assertEqualsForProto(ByteString.EMPTY, proto.getRepeatedBytes(1));
    assertEqualsForProto(
        ByteString.copyFrom([1, 2, 3]), proto.getRepeatedBytes(2));
    assertEqualsForProto(
        ByteString.copyFrom([4, 5, 6]), proto.getRepeatedBytes(3));
    assertEqualsForProto(
        ByteString.copyFrom([7, 8, 9]), proto.getRepeatedBytes(4));
  }

  testRepeatedField_set() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedBytes(TEST_STRING)
                           .addRepeatedBytes(ByteString.EMPTY)
                           .addRepeatedBytes(ByteString.copyFrom([0, 0, 0]))
                           .build();

    const builder = startProto.toBuilder().setRepeatedBytes(
        2, ByteString.copyFrom([3, 3, 3]));

    assertEqualsForProto(
        [
          TEST_STRING,
          ByteString.EMPTY,
          ByteString.copyFrom([3, 3, 3]),
        ],
        builder.getRepeatedBytesList().toArray());


    const proto = builder.build();


    assertEqualsForProto(
        [
          TEST_STRING,
          ByteString.EMPTY,
          ByteString.copyFrom([3, 3, 3]),
        ],
        proto.getRepeatedBytesList().toArray());
  }

  testRepeatedField_getAndCount() {
    const builder = TestProto.newBuilder()
                        .addRepeatedBytes(TEST_STRING)
                        .addRepeatedBytes(ByteString.EMPTY)
                        .addRepeatedBytes(ByteString.EMPTY);

    assertEqualsForProto(3, builder.getRepeatedBytesCount());
    assertEqualsForProto(TEST_STRING, builder.getRepeatedBytes(0));
    assertEqualsForProto(ByteString.EMPTY, builder.getRepeatedBytes(1));
    assertEqualsForProto(ByteString.EMPTY, builder.getRepeatedBytes(2));
    if (isCheckIndex()) {
      assertThrows(() => builder.getRepeatedBytes(3));
    }

    const proto = builder.build();
    assertEqualsForProto(3, proto.getRepeatedBytesCount());
    assertEqualsForProto(TEST_STRING, proto.getRepeatedBytes(0));
    assertEqualsForProto(ByteString.EMPTY, proto.getRepeatedBytes(1));
    assertEqualsForProto(ByteString.EMPTY, proto.getRepeatedBytes(2));
    if (isCheckIndex()) {
      assertThrows(() => proto.getRepeatedBytes(3));
    }
  }

  testRepeatedField_clear() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedBytes(TEST_STRING)
                           .addRepeatedBytes(ByteString.EMPTY)
                           .addRepeatedBytes(ByteString.copyFrom([0, 0, 0]))
                           .build();

    const builder = startProto.toBuilder().clearRepeatedBytes();
    assertEqualsForProto(0, builder.getRepeatedBytesCount());
    assertEqualsForProto(0, builder.getRepeatedBytesList().size());

    const proto = builder.build();
    assertEqualsForProto(0, proto.getRepeatedBytesCount());
    assertEqualsForProto(0, proto.getRepeatedBytesList().size());
  }
}

testSuite(new ByteStringFieldsTest());
